Fixes for issues #10, #48, #50, #55, #56, #60, #61, #63.

scommisso opened the pull request:
1. Fixed "useExternalSorting" config property (#48)
2. Added "canSelectRows" conrfig property (#10)
3. Changed footer template to hide the "selected count" when
   multi-select mode is not enabled ("isMultiSelect" = false, or
   "canSelectRows" = false)
4. Changed default "footerRowHeight" when multi-select mode is not
   enabled, and user has not overridden the default value.
5. Modified footer row CSS to align baselines of row count, page size
   label, page size select value, and pager value.
6. Fixed bug where "columnDefs" only renders columns if there are 2 or more columns.
7. Fixed bug where "rowHeight" was being used to build header css styles instead of "headerRowHeight"
8. Fixed sorting and filtering of complex properties.
9. Added support for disabling sorting on columns (e.g. for action columns like a delete link).
10. Added support for empty strings in colDef.displayName  (e.g. again, for action columns like a delete link)
11. Added @emrysx fixes for issue #63.

scommisso committed:
Fix issues #48 & #10

1. Fixed "useExternalSorting" config property (#48)
2. Added "canSelectRows" conrfig property (#10)
3. Changed footer template to hide the "selected count" when
multi-select mode is not enabled ("isMultiSelect" = false, or
"canSelectRows" = false)
4. Changed default "footerRowHeight" when multi-select mode is not
enabled, and user has not overridden the default value.
5. Modified footer row CSS to align baselines of row count, page size
label, page size select value, and pager value.

scommisso committed:
fixed #50

Fix for #50 - added support for columnDefs array with only 1 column.

scommisso committed:
Fixed bug in #10 implementation where clicks were not bubbling correctly

Fixed bug in #10 implementation where clicks were not bubbling correctly

scommisso committed:
Fixed #56

Fixed "headerRowHeight" configuration item, issue #56.

scommisso committed:
Fixed issue #55

Fixed sorting/filtering of complex properties.

scommisso committed:
Fixed #60, #61

1. Added support for disabling sorting on columns (e.g. action columns
like a delete link).
2. Added support for empty strings in colDef.displayName  (e.g. again,
for action columns like a delete link)

scommisso committed:
Fixed formatting

Converted CR\LF line breaks to LF line breaks.

scommisso committed:
Fixed formatting in build outputs

scommisso committed:
Updated implementation of #55

Modified binding path traversal of complex properties to stop when it
reaches a null or undefined value.

scommisso committed:
Fixed formatting in build files again...

Running the build batch file on a Windows machine adds CR/LF at the end
of the file -- after a build I need to use Notepad++ to edit the line
endings.

scommisso committed:
Cleaned up commented-out code from development...

scommisso commented on koGrid.debug.js:
```diff
@@ -133,17 +133,17 @@ kg.utils = utils;
     utils.forEach(cols, function (col, i) {
         if (col.field === '__kg_selected__') {
-            b.append('<div class="kgSelectionCell" data-bind="kgHeader: { value: \'{0}\' } ">', col.field);
+            b.append('<div class="kgSelectionCell" data-bind="kgHeader: { value: \'{0}\' }, css: { \'kgNoSort\': {1} }">', col.field, !col.allowSort);
```
The kgNoSort css class turns the cursor back to the default when hovering over a column header that does not support sorting.

emrysx replied:
The issue I just opened on Eric's original repo (https://github.com/ericmbarnard/KoGrid/issues/63) bites your changes here, and anywhere else you're using KO's "css" data-bind.

For your fork, changes to fix are:
ln 2714: element.className += " kgCell " + "col" + cell.column.index + " ";
ln 2814: element.className += " kgHeaderCell col" + cell.colIndex + " ";

scommisso replied:
Thanks! This has been fixed.

scommisso commented on koGrid.debug.js:
```diff
@@ -158,9 +158,9 @@ kg.utils = utils;
+    b.append('<div data-bind="click: $data.sort, css: { \'kgSorted\': !$data.noSortVisible() }">');
     b.append('  <span data-bind="text: $data.displayName"></span>');
-    b.append('  <img class="kgSortImg" data-bind="visible: $data.noSortVisible" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAcAAAAJCAYAAAD+WDajAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAadEVYdFNvZnR3YXJlAFBhaW50Lk5FVCB2My41LjEwMPRyoQAAAEFJREFUKFNjYICC+vp6YyDeDcSCMDEwDRRQAuK7QPwfpAAuiSYBkkQoAHLOQAVgEjB6FYrxGBy8OvHaide1+PwJAMBIWUlZ9vlNAAAAAElFTkSuQmCC"/>');
+    b.append('  <img class="kgSortImg" data-bind="visible: $data.allowSort() && $data.noSortVisible()" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAcAAAAJCAYAAAD+WDajAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAadEVYdFNvZnR3YXJlAFBhaW50Lk5FVCB2My41LjEwMPRyoQAAAEFJREFUKFNjYICC+vp6YyDeDcSCMDEwDRRQAuK7QPwfpAAuiSYBkkQoAHLOQAVgEjB6FYrxGBy8OvHaide1+PwJAMBIWUlZ9vlNAAAAA ...
```
Checking for allowSort() here prevents the default bi-directional arrow from showing on columns that do not allow sorting. This check is not required on the other two images as the column will not be sorted if sorting is not allowed. This will also allow the asc/desc arrows to show up if the column is sorted on at initialization, but the user is not allowed to change the sort direction.

scommisso commented on koGrid.debug.js:
```diff
@@ -179,11 +179,11 @@ kg.utils = utils;
         cols = options.columns;
 
-    b.append('<div data-bind="kgRow: $data, click: $data.toggleSelected, css: { kgSelected: $data.selected }">');
+    b.append('<div data-bind="kgRow: $data, click: $data.toggleSelected, css: { \'kgSelected\': $data.selected }">');
```
Fixed for older versions of IE -- css classes need to be in quotes.

scommisso commented on koGrid.debug.js:
```diff
@@ -225,16 +225,16 @@ kg.utils = utils;
 ﻿kg.templates.defaultFooterTemplate = function () {
     return '<div class="kgTotalSelectContainer" data-bind="visible: footerVisible">' +
-                '<div class="kgFooterTotalItems">' +
+                '<div class="kgFooterTotalItems" data-bind="css: {\'kgNoMultiSelect\': !isMultiSelect()}" >' +
```
Changed template to not show "Selected Items:" if the grid is not in multi select mode.

scommisso commented on koGrid.debug.js:
```diff
@@ -367,14 +367,21 @@ kg.utils = utils;
 
     this.field = colDef.field;
-    this.displayName = colDef.displayName || colDef.field;
+    if (colDef.displayName === undefined || colDef.displayName === null) {
```
Changed to allow empty strings in the displayName property. It will still default to field name if displayName is null or undefined.

scommisso commented on koGrid.debug.js:
```diff
@@ -367,14 +367,21 @@ kg.utils = utils;
 
     //sorting
-    this.allowSort = true;
+    if (colDef.sortable === undefined || colDef.sortable === null) {
```
Added optional "sortable" property to column definitions. Default value is true to keep backwards-compatibility. To turn off sorting on a column, users must explicitly turn this property to "false".

emrysx replied:
Oddly, setting sortable to false for a column is causing that col header to relocate to the first col (where filter icon would normally be). Haven't figured out why yet.

emrysx replied:
Nm, figured it out. Is the class issue I noted in my comments further up.

emrysx replied:
By the way, thanks for adding sortable to cols. Exactly what I needed for an action column. Now if we can just get Eric to pull in your changes and keep this going.

scommisso replied:
Yeah he's been inactive for a while -- there are a couple of other pull requests pending as well.

scommisso commented on koGrid.debug.js:
```diff
@@ -424,8 +431,9 @@ kg.ColumnCollection.fn = {
-kg.Row = function (entity) {
+kg.Row = function (entity, config) {
     var self = this;
+    var canSelectRows = config.canSelectRows;
```
Added "canSelectRows" property to determine whether or not selection is even available in the grid. If this value is "false", "isMultiSelect" will have no effect. Default value is "true" (assigned in the grid initialization)

scommisso commented on koGrid.debug.js:
```diff
@@ -434,16 +442,25 @@ kg.Row = function (entity) {
 
     this.selected = ko.dependentObservable({
         read: function () {
+            if (!canSelectRows) {
```
Reference whether or not selection is enabled before checking/writing to selection observables.

scommisso commented on koGrid.debug.js:
```diff
@@ -550,6 +567,8 @@ kg.Row = function (entity) {
 
     this.filterVisible = ko.observable(false);
 
+    this.allowSort = ko.observable(col.allowSort);
```
Added observable to control whether or not a column is able to be sorted on.

scommisso commented on koGrid.debug.js:
```diff
@@ -642,7 +664,7 @@ kg.Row = function (entity) {
 
             // build the row
-            row = new kg.Row(entity);
+            row = new kg.Row(entity, grid.config);
```
Modified Row constructor to pass the config down the chain so that it can reference whether or not a row can be selected.

scommisso commented on koGrid.debug.js:
```diff
@@ -938,13 +963,17 @@ kg.Row = function (entity) {
 
                     // pull the data out of the item
-                    itemData = ko.utils.unwrapObservable(item[field]);
+                    propPath = col.field.split(".");
```
Support filtering on complex properties.

emrysx replied:
Line 966 should be: propPath = field.split(".");
Filtering was broken for me till I fixed this.

scommisso replied:
Sorry, copy-paste error and I wasn't testing with filtering turned on... this has been fixed, thanks!

scommisso commented on koGrid.debug.js:
```diff
@@ -1269,7 +1300,11 @@ kg.Row = function (entity) {
 
             if (item) {
-                prop = ko.utils.unwrapObservable(item[col.field]);
+                propPath = col.field.split(".");
```
Support sorting on complex properties.

scommisso commented on koGrid.debug.js:
```diff
@@ -1281,16 +1316,26 @@ kg.Row = function (entity) {
+                propPath,
+                i;
+
+            propPath = col.field.split(".");
```
Support sorting on complex properties.

scommisso commented on koGrid.debug.js:
```diff
@@ -1676,6 +1722,14 @@ kg.KoGrid = function (options) {
     this.initPhase = 0;
 
 
+    // Set new default footer height if not overridden, and multi select is disabled
```
Adjust default footer height if the "Selected Items:" text has been hidden (e.g. Grid is not in multi-select mode) -- but only if the user of the grid has not specified a custom footerRowHeight.

scommisso commented on koGrid.debug.js:
```diff
@@ -1684,7 +1738,7 @@ kg.KoGrid = function (options) {
         data: filterManager.filteredData,
         sortInfo: self.config.sortInfo,
-        useExternalSorting: self.config.useExternalFiltering
+        useExternalSorting: self.config.useExternalSorting
```
Fix bug where externalSorting could never be turned on (referencing wrong config property)

scommisso commented on koGrid.debug.js:
```diff
@@ -2003,7 +2057,7 @@ kg.KoGrid = function (options) {
         }
 
-        if (columnDefs.length > 1) {
+        if (columnDefs.length > 0) {
```
Fixed a bug where a grid with only 1 column would never initialize properly.

scommisso commented on koGrid.debug.js:
```diff
@@ -2111,6 +2165,7 @@ kg.cssBuilder = {
 
     buildStyles: function (grid) {
         var rowHeight = (grid.config.rowHeight - grid.elementDims.rowHdiff),
+            headerRowHeight = grid.config.headerRowHeight,
```
Fixed header row to correctly reference "headerRowHeight" config property.

scommisso commented on koGrid.debug.js:
```diff
@@ -2605,6 +2660,9 @@ ko.bindingHandlers['kgRow'] = (function () {
                 rowManager = bindingContext.$parent.rowManager,
                 rowSubscription;
 
+            if (grid.config.canSelectRows) {
```
Added kgCanSelect css class to control CSS. If a row is not selectable, cursor will be default. If row is selectable (has kgCanSelect class), cursor will be pointer.

scommisso committed:
Empty cellClass value causes problems with custom css binding.

If the cellClass value is empty, adding custom css bindings in a
cellTemplate breaks the css rendering. For simplicity of fix, if
cellClass is empty, a "kgEmpty" class is applied (note: there is no
matching rule for this, so there is no change to the rendering)

scommisso committed:
fixed line ending formatting from last checkin

thelinuxlich commented:
wonderful!

scommisso committed:
Incorporate changes provided by @emrysx

Incorporate changes provided by @emrysx. See comments in pull request 49
for details.

scommisso committed:
Fixed IE9 rendering of last page button
